custodia/ipa/certreq.py

```diff
@@ -0,0 +1,257 @@
+# Copyright (C) 2017  Custodia Project Contributors - see LICENSE file
+"""FreeIPA cert request store
+"""
+import datetime
+import textwrap
+
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.asymmetric import rsa
+from cryptography.x509 import oid
+
+from ipalib.krb_utils import krb5_format_service_principal_name
+
+import six
+
+from custodia.plugin import CSStore, CSStoreError, PluginOption, REQUIRED
+from .interface import IPAInterface
+
+
+TLS_SERVERAUTH = oid.ObjectIdentifier('2.5.29.37.1')
+
+TEMPLATE = textwrap.dedent("""\
+    Issuer: {issuer}
+    Subject: {subject}
+    Serial Number: {cert.serial_number}
+    Validity:
+        Not Before: {cert.not_valid_before}
+        Not After: {cert.not_valid_after}
+    {pem}\
+""")
+
+
+class CSRBuilder(object):
+    def __init__(self, plugin, backend=None):
+        if backend is None:
+            self.backend = default_backend()
+        else:
+            self.backend = backend
+        self.plugin = plugin
+
+    def gen_private(self):
+        return rsa.generate_private_key(
+            public_exponent=65537,
```

Comment:
I'd find it more confusing. Any person with small exposure to crypto and math will sooner or later get familiar with 65537.

Fermat prime is (mostly) a red herring. In theory any prime will do. People have used **3** in the past. In practice the size of 65537 is the best compromise between performance and security. A Fermat prime has the additional benefit, https://en.wikipedia.org/wiki/65537_(number) 

```
>>> 2**2**4 + 1
65537
>>> bin(2**2**4 + 1)
'0b10000000000000001'
```